### PR TITLE
CLOUDSTACK-10434:Some APIs should have access check

### PR DESCRIPTION
**Pull request by lujiefsi**
### Description

This PR fixes the CLOUDSTACK-10434. I think some APIs lack access check and  list them in below table. I also give the pattch to add the access check for the api in this table. Anyone chould change this table, If you think the APIs do not need access check and change their lable as "no".

|  API   | Lack?  |
|  ----  | ----  |
| VolumeApiServiceImpl # updateVolume  | yes |
| VolumeApiServiceImpl # detachVolumeViaDestroyVM  | yes |
| VolumeApiServiceImpl # takeSnapshot  | yes |
| VolumeApiServiceImpl # migrateVolume  | yes |
| AccountManagerImpl#createApiKeyAndSecretKey | yes |
| LoadBalancingRulesManagerImpl#applyLBStickinessPolicy | yes |
| LoadBalancingRulesManagerImpl#applyLBHealthCheckPolicy| yes |
| TemplateManagerImpl#createPrivateTemplate| yes |
| SnapshotManagerImpl#updateSnapshotPolicy| yes |

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [x] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [x] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Review by DaanHoogland — APPROVED** [review]
code looks good

**Comment by lujiefsi** [comment]
my local test is good 

[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  10:43 min
[INFO] Finished at: 2021-03-24T17:30:51+08:00
[INFO] ------------------------------------------------------------------------

**Comment by lujiefsi** [comment]
run the test again locallly, and everything is ok.
`[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  10:02 min
[INFO] Finished at: 2021-03-25T16:53:40+08:00
[INFO] ------------------------------------------------------------------------`

It's strange. Anyway, submit an empty commit to trigger test again.

**Comment by DaanHoogland** [comment]
> It's strange. Anyway, submit an empty commit to trigger test again.

@lujiefsi you can do this by closing and reopening on github, you don't have to push.

**Comment by lujiefsi** [comment]
> > It's strange. Anyway, submit an empty commit to trigger test again.
> 
> @lujiefsi you can do this by closing and reopening on github, you don't have to push.

Got it.  I have merged these commits as one.

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by Pearl1594** [comment]
@blueorangutan test

**Comment by lujiefsi** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan test centos7 vmware-67u3

**Comment by rhtyd** [comment]
@Spaceman1984 can you manually test those APIs, I reckon those are not covered by smoketests

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java [hunk 1]
```diff
@@ -2199,6 +2207,7 @@ public Volume migrateVolume(MigrateVolumeCmd cmd) {
             throw new InvalidParameterValueException("Volume " + vol + " is already on the destination storage pool");
         }
 
+        _accountMgr.checkAccess(caller, null, true, vol);
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java:2210]: can call this as soon as `vol` is not null (at line #2201 ^^^)

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java [hunk 2]
```diff
@@ -2556,7 +2566,7 @@ private Snapshot takeSnapshotInternal(Long volumeId, Long policyId, Long snapsho
         if (volume.getState() != Volume.State.Ready) {
             throw new InvalidParameterValueException("VolumeId: " + volumeId + " is not in " + Volume.State.Ready + " state but " + volume.getState() + ". Cannot take snapshot.");
         }
-
+        _accountMgr.checkAccess(caller, null, true, volume);
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java:2569]: it's better to call this before checking for the volume state.

File: server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java [hunk 3]
```diff
@@ -335,6 +335,10 @@ public SnapshotPolicy updateSnapshotPolicy(UpdateSnapshotPolicyCmd cmd) {
         Boolean display = cmd.getDisplay();
 
         SnapshotPolicyVO policyVO = _snapshotPolicyDao.findById(id);
+        VolumeInfo volume = volFactory.getVolume(policyVO.getVolumeId());
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java:338]: there are chances of returning 'null' for VolumeInfo here.

File: server/src/main/java/com/cloud/template/TemplateManagerImpl.java [hunk 4]
```diff
@@ -1687,6 +1689,7 @@ public VirtualMachineTemplate createPrivateTemplate(CreateTemplateCmd command) t
             } else if (volumeId != null) {
                 VolumeInfo volInfo = _volFactory.getVolume(volumeId);
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/template/TemplateManagerImpl.java:1691]: `volInfo` can be null in some cases. please check.

**Comment by rhtyd** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java [hunk 5]
```diff
@@ -1994,6 +1997,10 @@ private void validateRootVolumeDetachAttach(VolumeVO volume, UserVmVO vm) {
 
     @ActionEvent(eventType = EventTypes.EVENT_VOLUME_DETACH, eventDescription = "detaching volume")
     public Volume detachVolumeViaDestroyVM(long vmId, long volumeId) {
+        Account caller = CallContext.current().getCallingAccount();
+        Volume volume = _volsDao.findById(volumeId);
+        // Permissions check
+        _accountMgr.checkAccess(caller, null, true, volume);
```
Comment (sureshanaparti) [inline server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java:2003]: @lujiefsi VM destroy might fail if the caller cannot access any volume(s) attached to VM. So, I think it is better to check the access before destroying VM.

Comment (lujiefsi) [reply server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java:2003]: destroyVolume alerady have an access check, in method "retrieveAndValidateVolume"

Comment (lujiefsi) [reply server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java:2003]: detachVolumeFromVM still needs the access check to precent others detach a volume that does not belong to him

**Comment by rhtyd** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java (hunk 5 above)
Comment (sureshanaparti) [reply server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java:2003]: @lujiefsi I mean, access check for all the volumes before destroying VM.

Comment (lujiefsi) [reply server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java:2003]: @sureshanaparti  got it and see UserVmManagerImpl

**Comment by nvazquez** [comment]
Hi @lujiefsi can you please resolve the conflicts?

**Comment by rhtyd** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 6]
```diff
@@ -2966,9 +2966,20 @@ public UserVm destroyVm(DestroyVMCmd cmd) throws ResourceUnavailableException, C
         checkForUnattachedVolumes(vmId, volumesToBeDeleted);
         validateVolumes(volumesToBeDeleted);
 
+        List<VolumeVO> allVolumes = null;
+        if (vm.getHypervisorType() == HypervisorType.VMware) {
+            allVolumes = _volsDao.findByInstance(vm.getId());
+            allVolumes.removeIf(vol -> vol.getVolumeType() == Volume.Type.ROOT);
+        }else {
+            allVolumes = volumesToBeDeleted;
+        }
```
Comment (Pearl1594) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:2975]: @lujiefsi I believe this section isn't required as it causes issues wrt recovery of destroyed VMs with their data disks on VMWare (https://github.com/apache/cloudstack/pull/4493/files#diff-865f710fe60b9b62ff8c2cbce8a8342e4f6fca658223119762285fd5f7857f08L2949).

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 7]
```diff
@@ -2966,9 +2966,20 @@ public UserVm destroyVm(DestroyVMCmd cmd) throws ResourceUnavailableException, C
...
+        for (VolumeVO volume : allVolumes) {
+             _accountMgr.checkAccess(ctx.getCallingAccount(), null, true, volume);
+        }
+
```
Comment (Pearl1594) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:2979]: We could probably use the following to check access for all volumes to be deleted?
```suggestion
       final ControlledEntity[] volumesToDelete = volumesToBeDeleted.toArray(new ControlledEntity[0]);
        _accountMgr.checkAccess(ctx.getCallingAccount(), null, true, volumesToDelete);

```

**Comment by rhtyd** [comment]
@lujiefsi can you address review comments?
@Spaceman1984 ping - any update on manual testing?

**Comment by lujiefsi** [comment]
> @lujiefsi can you address review comments?
> @Spaceman1984 ping - any update on manual testing?

YES, i am addessing review comments and  testing it.

**Comment by Spaceman1984** [comment]
Tested all calls manually except LoadBalancingRulesManagerImpl#applyLBHealthCheckPolicy, I'm not sure how to. @rhtyd.
Seems like I need Netscaler infra.

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java (hunk 6 above)
Comment (lujiefsi) [reply server/src/main/java/com/cloud/vm/UserVmManagerImpl.java]: ops, have fixed it

**Comment by Pearl1594** [comment]
@blueorangutan package

**Review by nvazquez — APPROVED** [review]
Code LGTM

**Comment by Pearl1594** [comment]
@blueorangutan test

**Review by Pearl1594 — APPROVED** [review]
Code LGTM. Did basic tests, works as expected.

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java (hunk 5 above)
Comment (sureshanaparti) [reply server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java:2022]: thanks @lujiefsi 
```suggestion
        _accountMgr.checkAccess(caller, null, true, volume);
```

**Review by sureshanaparti — APPROVED** [review]

**Comment by Pearl1594** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by Spaceman1984** [comment]
> Tested all calls manually except LoadBalancingRulesManagerImpl#applyLBHealthCheckPolicy, I'm not sure how to. @rhtyd.
> Seems like I need Netscaler infra.
> 
> Other than that, no issues found.

I picked up an issue with one of the calls, it doesn't break anything, I just feel that the check is not needed in this particular situation. In AccountManagerImpl#createApiKeyAndSecretKey - this method is called the first time the management server starts up and the baremetal system user doesn't exist. The checkAccess method then checks if the system user is allowed to access the baremetal system user account, which always returns true for the system account.

API Command Name | Method | Owner | Different User 
----|----|----|----
updateVolume | VolumeApiServiceImpl # updateVolume | Pass | Pass 
destroyVirtualMachine | VolumeApiServiceImpl # detachVolumeViaDestroyVM | Pass | Pass
createSnapshot | VolumeApiServiceImpl # takeSnapshot | Pass | Pass 
migrateVolume | VolumeApiServiceImpl # migrateVolume | Pass | Pass
N/A - called to create baremetal user the first time the management server starts up. - I don't believe we need this check| AccountManagerImpl#createApiKeyAndSecretKey | N/A | N/A
createLBStickinessPolicy | LoadBalancingRulesManagerImpl#applyLBStickinessPolicy | Pass | Pass 
createLBHealthCheckPolicy | LoadBalancingRulesManagerImpl#applyLBHealthCheckPolicy | skipped | skipped
createTemplate | TemplateManagerImpl#createPrivateTemplate | Pass | Pass
updateSnapshotPolicy | SnapshotManagerImpl#updateSnapshotPolicy | Pass | Pass

**Comment by rhtyd** [comment]
Merging based on @Spaceman1984 's test report, smoke tests and reviews.

**Pull request merged**